extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/dev/HibernateOrmDevJsonRpcService.java

```diff
@@ -18,4 +46,100 @@ public int getNumberOfNamedQueries() {
         return getInfo().getNumberOfNamedQueries();
     }
 
+    private Optional<HibernateOrmDevInfo.PersistenceUnit> findPersistenceUnit(String persistenceUnitName) {
+        return getInfo().getPersistenceUnits().stream().filter(pu -> pu.getName().equals(persistenceUnitName)).findFirst();
+    }
+
+    public DataSet executeHQL(String persistenceUnit, String hql, Integer pageNumber, Integer pageSize) {
+        if (!isDev) {
+            return errorDataSet("This method is only allowed in dev mode");
+        }
+
+        if (!hqlIsValid(hql)) {
+            return errorDataSet("The provided HQL was not valid");
+        }
+
+        Optional<HibernateOrmDevInfo.PersistenceUnit> pu = findPersistenceUnit(persistenceUnit);
+        if (pu.isEmpty()) {
+            return errorDataSet("No such persistence unit: " + persistenceUnit);
+        }
+
+        //noinspection resource
+        SessionFactoryImplementor sf = pu.get().sessionFactory();
+
+        // Check the connection for this persistence unit points to an allowed datasource
+        ConnectionProvider connectionProvider = sf.getServiceRegistry().requireService(ConnectionProvider.class);
+        if (connectionProvider instanceof QuarkusConnectionProvider quarkusConnectionProvider) {
+            if (!isAllowedDatabase(quarkusConnectionProvider.getDataSource())) {
+                return errorDataSet("The persistence unit datasource points to a non-allowed datasource " +
+                        "(by default, only local databases are allowed).");
+            }
+        } else {
+            return errorDataSet("Unsupported Connection Provider type for specified persistence unit.");
+        }
```

Comment:
LGTM.

Maybe you could add (and document) a special value (`*`?) for the `allowed-host` property, so that these two error cases can be avoided when people are fine with allowing all hosts? Since that'd be opt-in, it'd still be relatively safe.

Reply:
Done (for both the Hibernate and Agroal Dev UI back-ends).